### PR TITLE
Rework of MachInst isel, branch fixups and lowering, and block ordering.

### PR DESCRIPTION
**Pull request by cfallin**
tl;dr: new new-isel; better block-ordering, handling branches in one pass. 24% faster compile+run on bz2 (28% fewer instructions); 10% faster compile (10% fewer instructions).

This patch includes:

- A complete rework of the way that CLIF blocks and edge blocks are
  lowered into VCode blocks. The new mechanism in `BlockLoweringOrder`
  computes RPO over the CFG, but with a twist: it merges edge blocks intto
  heads or tails of original CLIF blocks wherever possible, and it does
  this without ever actually materializing the full nodes-plus-edges
  graph first. The backend driver lowers blocks in final order so
  there's no need to reshuffle later.

- A new `MachBuffer` that replaces the `MachSection`. This is a special
  version of a code-sink that is far more than a humble `Vec<u8>`. In
  particular, it keeps a record of label definitions and label uses,
  with a machine-pluggable `LabelUse` trait that defines various types
  of fixups (basically internal relocations).

  Importantly, it implements some simple peephole-style branch rewrites
  *inline in the emission pass*, without any separate traversals over
  the code to use fallthroughs, swap taken/not-taken arms, etc. It
  tracks branches at the tail of the buffer and can (i) remove blocks
  that are just unconditional branches (by redirecting the label), (ii)
  understand a conditional/unconditional pair and swap the conditional
  polarity when it's helpful; and (iii) remove branches that branch to
  the fallthrough PC.

  The `MachBuffer` also implements branch-island support. On
  architectures like AArch64, this is needed to allow conditional
  branches within plausibly-attainable ranges (+/- 1MB on AArch64
  specifically). It also does this inline while streaming through the
  emission, without any sort of fixpoint algorithm or later moving of
  code, by simply tracking outstanding references and "deadlines" and
  emitting an island just-in-time when we're in danger of going out of
  range.

- A rework of the instruction selector driver. This is largely following
  the same algorithm as before, but is cleaned up significantly, in
  particular in the API: the machine backend can ask for an input arg
  and get any of three forms (constant, register, producing
  instruction), indicating it needs the register or can merge the
  constant or producing instruction as appropriate. This new driver
  takes special care to emit constants right at use-sites (and at phi
  inputs), minimizing their live-ranges, and also special-cases the
  "pinned register" to avoid superfluous moves.

Overall, on `bz2.wasm`, the results are:

```
    wasmtime full run (compile + runtime) of bz2:

    baseline:   9774M insns, 9742M cycles, 3.918s
    w/ changes: 7012M insns, 6888M cycles, 2.958s  (24.5% faster, 28.3% fewer insns)

    clif-util wasm compile bz2:

    baseline:   2633M insns, 3278M cycles, 1.034s
    w/ changes: 2366M insns, 2920M cycles, 0.923s  (10.7% faster, 10.1% fewer insns)

    All numbers are averages of two runs on an Ampere eMAG.
```

### CODE REVIEW

**Comment by cfallin** [comment]
Updated: rebased onto latest master; brought x64 backend up-to-date; updated all aarch64 filetests; and added some more tests for MachBuffer. Should be ready for review now.

**Comment by bjorn3** [comment]
> I see debug! calls in potentially hottish places, eg MachBuffer::put1(). Are we sure those become zero cost in release builds?

`trace!` is probably more appropriate.

File: cranelift/codegen/src/isa/x64/inst/args.rs [hunk 1]
```diff
@@ -421,31 +404,17 @@ impl BranchTarget {
...
-        }
-    }
-
-    /// Map the block index given a transform map.
-    pub fn map(&mut self, block_index_map: &[BlockIndex]) {
-        match self {
-            &mut BranchTarget::Block(ref mut bix) => {
-                let n = block_index_map[*bix as usize];
-                *bix = n;
+                assert!(off >= -0x7FFF_FF00 && off <= 0x7FFF_FF00);
```
Comment (julian-seward1) [inline cranelift/codegen/src/isa/x64/inst/args.rs:414]: This change concerns me somewhat.  What guarantees that this assertion can't fail now?

File: cranelift/codegen/src/isa/x64/inst/mod.rs [hunk 2]
```diff
@@ -868,87 +811,95 @@ impl MachInst for Inst {
...
 
-    fn gen_jump(blockindex: BlockIndex) -> Inst {
-        Inst::jmp_known(BranchTarget::Block(blockindex))
+    fn gen_jump(label: MachLabel) -> Inst {
+        Inst::jmp_known(BranchTarget::Label(label))
+    }
+
+    fn gen_constant(to_reg: Writable<Reg>, value: u64, _: Type) -> SmallVec<[Self; 4]> {
+        let mut ret = SmallVec::new();
+        let is64 = value > 0xffff_ffff;
```
Comment (julian-seward1) [inline cranelift/codegen/src/isa/x64/inst/mod.rs:820]: Is this definition of `is64` right?  That seems like it's an unsigned criterion, but the general rule on Intel for 32-bit immediate fields is that they are sign extended to 64 bits as appropriate.  If this logic simply moved from elsewhere in this patch, then leave it as is; but otherwise maybe change to the signed variant, using `low32willSXto64` ?

File: cranelift/codegen/src/isa/x64/lower.rs [hunk 3]
```diff
@@ -219,35 +230,6 @@ fn lower_insn_to_regs<'a>(ctx: Ctx<'a>, iri: IRInst) {
             panic!("ALU+imm and ALU+carry ops should not appear here!");
         }
 
-        Opcode::X86Udivmodx
```
Comment (julian-seward1) [inline cranelift/codegen/src/isa/x64/lower.rs:222]: I feel like it's a shame to lose this, because it means losing the ability to easily differentiate legitimate failures due to non-implementation of a target-independent CLIR insn, vs bugs resulting in machine-specific CLIRs being handed to us.

Comment (cfallin) [reply cranelift/codegen/src/isa/x64/lower.rs:222]: Oh, these `X86*` opcodes went away in the latest `master`, so this is just a rebase-related change. I think the rule is still (as has always been) never have a fallthrough in the big-opcode-match, and handle additions or deletions as they come, indicated by compile errors -- so if any machine-specific ops are added back in the future, we'll figure out what to do then.

File: cranelift/codegen/src/isa/x64/inst/args.rs (hunk 1 above)
Comment (cfallin) [reply cranelift/codegen/src/isa/x64/inst/args.rs:414]: Nothing guarantees it, but this form (`ResolvedOffset`) is used only when the lowering explicitly selects it, now; ordinary branches don't go through this code (the `LabelUse` handles them instead, and we can implement 64-bit long-form veneers there if we think we need >2GB code-size).

File: cranelift/codegen/src/isa/x64/inst/mod.rs (hunk 2 above)
Comment (cfallin) [reply cranelift/codegen/src/isa/x64/inst/mod.rs:820]: I think so, or at least, this was existing logic in the `Iconst` lowering prior to this patch:

https://github.com/bytecodealliance/wasmtime/blob/a75377565f830052094aa8aa72c5e7a6b787fa18/cranelift/codegen/src/isa/x64/lower.rs#L116

File: cranelift/codegen/src/machinst/lower.rs [hunk 4]
```diff
@@ -731,27 +848,94 @@ impl<'func, I: VCodeInst> LowerCtx for Lower<'func, I> {
...
+        let val = self.f.dfg.resolve_aliases(val);
+        self.get_input_for_val(ir_inst, val)
+    }
+
+    fn get_output(&mut self, ir_inst: Inst, idx: usize) -> Writable<Reg> {
+        let val = self.f.dfg.inst_results(ir_inst)[idx];
+        Writable::from_reg(self.value_regs[val])
+    }
+
+    fn tmp(&mut self, rc: RegClass, ty: Type) -> Writable<Reg> {
```
[the remark this answers is not in the document]
Comment (cfallin) [reply cranelift/codegen/src/machinst/lower.rs]: Was part of the old API, but no reason not to rename here :-) Now it's `alloc_tmp()`.

File: cranelift/codegen/src/machinst/lower.rs [hunk 5]
```diff
@@ -229,452 +373,427 @@ impl<'func, I: VCodeInst> Lower<'func, I> {
...
-        self.f.dfg.inst_args(ir_inst).len()
-    }
+        // Now that we've emitted all instructions into the VCodeBuilder, let's build the VCode.
+        let vcode = self.vcode.build();
+        debug!("built vcode: {:?}", vcode);
+
+        Ok(vcode)
+    }
+
+    fn get_input_for_val(&self, at_inst: Inst, val: Value) -> LowerInput {
```
[the remark this answers is not in the document]
Comment (cfallin) [reply cranelift/codegen/src/machinst/lower.rs:732]: Done!

File: cranelift/codegen/src/machinst/lower.rs [hunk 6]
```diff
@@ -3,104 +3,191 @@
...
 use alloc::boxed::Box;
 use alloc::vec::Vec;
 use log::debug;
 use smallvec::SmallVec;
-use std::collections::VecDeque;
 
-/// A context that machine-specific lowering code can use to emit lowered instructions. This is the
-/// view of the machine-independent per-function lowering context that is seen by the machine
-/// backend.
+/// An "instruction color" partitions instructions by side-effecting ops. All
```
[the remark this answers is not in the document]
Comment (cfallin) [reply cranelift/codegen/src/machinst/lower.rs]: Done.

File: cranelift/codegen/src/machinst/lower.rs [hunk 7]
```diff
@@ -3,104 +3,191 @@
...
+/// instructions with the same "color" are guaranteed not to be separated by any
+/// side-effecting op (for this purpose, loads are also considered
+/// side-effecting, to avoid subtle questions w.r.t. the memory model), and
+/// furthermore, it is guaranteed that for any two instructions A and B such
+/// that color(A) == color(B), either A dominates B and B postdominates A, or
+/// vice-versa. (For now, in practice, only ops in the same basic block can ever
+/// have the same color, trivially providing the second condition.) Intuitively,
+/// this means that the ops of the same color must always execute "together", as
+/// part of one atomic contiguous section of the dynamic execution trace, and
+/// they can be freely permuted without affecting program behavior.
```
[the remark this answers is not in the document]
Comment (cfallin) [reply cranelift/codegen/src/machinst/lower.rs]: Yep, modulo true deps; clarified.

File: cranelift/codegen/src/machinst/buffer.rs [hunk 8]
```diff
@@ -0,0 +1,1035 @@
...
+    /// Any call site records referring to this code.
+    call_sites: SmallVec<[MachCallSite; 16]>,
+    /// Any source location mappings referring to this code.
+    srclocs: SmallVec<[MachSrcLoc; 64]>,
+    /// The current source location in progress (after `start_srcloc()` and
+    /// before `end_srcloc()`).  This is a (start_offset, src_loc) tuple.
+    cur_srcloc: Option<(CodeOffset, SourceLoc)>,
+    /// Known label offsets; `UNKNOWN_LABEL_OFFSET` if unknown.
+    label_offsets: SmallVec<[CodeOffset; 16]>,
+    /// Label aliases: one label points to an unconditional jump to another
```
[the remark this answers is not in the document]
Comment (cfallin) [reply cranelift/codegen/src/machinst/buffer.rs]: Was correct but probably too terse; fixed.

File: cranelift/codegen/src/machinst/blockorder.rs [hunk 9]
```diff
@@ -1,49 +1,579 @@
 //! Computation of basic block order in emitted code.
+//!
+//! This module handles the translation from CLIF BBs to VCode BBs.
+//!
+//! The basic idea is that we compute a sequence of "lowered blocks" that
```
[the remark this answers is not in the document]
Comment (cfallin) [reply cranelift/codegen/src/machinst/blockorder.rs:5]: Hopefully the ASCII art and additional explanation help! This is pretty subtle so I'm happy to clarify further if needed.

File: cranelift/codegen/src/isa/aarch64/lower_inst.rs [hunk 10]
```diff
@@ -2072,6 +2082,11 @@ pub(crate) fn lower_branch<C: LowerCtx<I = Inst>>(
                 //   [jumptable offsets relative to JT base]
                 let jt_size = targets.len() - 1;
                 assert!(jt_size <= std::u32::MAX as usize);
+
+                ctx.emit(Inst::EmitIsland {
```
[the remark this answers is not in the document]
Comment (cfallin) [reply cranelift/codegen/src/isa/aarch64/lower_inst.rs:2088]: Added some more docs on `EmitIsland` to clarify.

File: cranelift/codegen/src/isa/aarch64/inst/mod.rs [hunk 11]
```diff
@@ -2683,15 +2577,134 @@ impl ShowWithRRU for Inst {
...
+    /// 26-bit branch offset (unconditional branches). PC-rel, offset is imm << 2. Immediate is 26
+    /// signed bits, in bits 25:0. Used by b, bl.
+    Branch26,
+    /// 19-bit offset for LDR (load literal). PC-rel, offset is imm << 2. Immediate is 19 signed bits,
+    /// in bits 23:5.
+    Ldr19,
+    /// 21-bit offset for ADR (get address of label). PC-rel, offset is not shifted. Immediate is
+    /// 21 signed bits, with high 19 bits in bits 23:5 and low 2 bits in bits 30:29.
+    Adr21,
+    /// 32-bit PC relative constant offset (from address of constant itself). Used in jump tables.
```
[the remark this answers is not in the document]
Comment (cfallin) [reply cranelift/codegen/src/isa/aarch64/inst/mod.rs]: Signed (clarified).

File: cranelift/codegen/src/isa/aarch64/inst/mod.rs [hunk 12]
```diff
@@ -1815,101 +1800,25 @@ impl MachInst for Inst {
...
-                targets: ref mut t, ..
-            } => {
-                for target in t.iter_mut() {
-                    // offset+20: jumptable is 20 bytes into compound sequence.
-                    target.lower(targets, my_offset + 20);
-                }
-            }
-            _ => {}
-        }
+    fn worst_case_size() -> CodeOffset {
```
[the remark this answers is not in the document]
Comment (cfallin) [reply cranelift/codegen/src/isa/aarch64/inst/mod.rs:1843]: Good idea! I added a debug assert to `Inst::emit()` that verifies that no more than `worst_case_size()` bytes were emitted.

File: cranelift/codegen/src/isa/aarch64/inst/mod.rs [hunk 13]
```diff
@@ -1789,6 +1757,23 @@ impl MachInst for Inst {
...
+            ret.push(Inst::load_fp_constant64(to_reg, f64::from_bits(value)));
+            ret
+        } else if ty == F32 {
+            let mut ret = SmallVec::new();
+            ret.push(Inst::load_fp_constant32(
+                to_reg,
+                f32::from_bits(value as u32),
+            ));
+            ret
+        } else {
```
[the remark this answers is not in the document]
Comment (cfallin) [reply cranelift/codegen/src/isa/aarch64/inst/mod.rs:1798]: Added assert.

File: cranelift/codegen/src/isa/aarch64/inst/mod.rs [hunk 14]
```diff
@@ -732,21 +728,19 @@ pub enum Inst {
...
-
     /// Marker, no-op in generated code: SP "virtual offset" is adjusted. This
     /// controls MemArg::NominalSPOffset args are lowered.
     VirtualSPOffsetAdj {
         offset: i64,
     },
+
+    /// Meta-insn, no-op in generated code: emit constant/branch veneer island at this point (with
+    /// a guard jump around it) if less than the needed space is available before the next branch
+    /// deadline.
```
[the remark this answers is not in the document]
Comment (cfallin) [reply cranelift/codegen/src/isa/aarch64/inst/mod.rs]: Added to the top of `machinst/buffer.rs`.

File: cranelift/codegen/src/isa/aarch64/inst/mod.rs [hunk 15]
```diff
@@ -685,11 +678,14 @@ pub enum Inst {
         trap_info: (SourceLoc, TrapCode),
     },
 
-    /// Load the address (using a PC-relative offset) of a MemLabel, using the
-    /// `ADR` instruction.
+    /// Load the address (using a PC-relative offset) of a memory location, using the `ADR`
```
[the remark this answers is not in the document]
Comment (cfallin) [reply cranelift/codegen/src/isa/aarch64/inst/mod.rs]: Done.

File: cranelift/codegen/src/isa/aarch64/inst/mod.rs [hunk 16]
```diff
@@ -645,35 +645,28 @@ pub enum Inst {
...
         taken: BranchTarget,
         not_taken: BranchTarget,
         kind: CondBrKind,
     },
 
-    /// Lowered conditional branch: contains the original branch kind (or the
-    /// inverse), but only one BranchTarget is retained. The other is
-    /// implicitly the next instruction, given the final basic-block layout.
-    CondBrLowered {
+    /// A one-way conditional branch, invisible to the CFG processing; used *only* as part of
```
[the remark this answers is not in the document]
Comment (cfallin) [reply cranelift/codegen/src/isa/aarch64/inst/mod.rs:658]: Clarified; this is the same as the original "lowered" form, but we just call out the actual semantics and purpose a bit more explicitly now rather than conflating it with the branch-lowering process.

File: cranelift/codegen/src/machinst/lower.rs [hunk 17]
```diff
@@ -3,104 +3,191 @@
...
+    /// - A register.
+    ///
+    /// The instruction input may be available in some or all of these
+    /// forms. More than one is possible: e.g., it may be produced by an
+    /// instruction in the same block, but may also have been forced into a
+    /// register already by an earlier op. It will *always* be available
+    /// in a register, at least.
+    ///
+    /// If the backend uses the register, rather than one of the other
+    /// forms (constant or merging of the producing op), it must call
```
[the remark this answers is not in the document]
Comment (cfallin) [reply cranelift/codegen/src/machinst/lower.rs:124]: Done.

**Comment by cfallin** [comment]
Thanks a bunch for the review! Hopefully I've addressed everything. A few responses to the questions below:

> There are a few comments in-line, but I have some larger semantics-level and maintainability questions here:
> 
>     * > The MachBuffer also implements branch-island support. [..]
>       
>       
>       
>       * How is this tested, considering it is low probability path stuff?  Given that long range jumps are only for > 1 MB on arm64, I feel like there's a bunch of paths here with very low probability of being taken.  Which is a verification hazard.  Is that the case?  If so, how is this stuff being tested?
>         (Later) I see tests at the bottom of machinst/buffer.rs.  Are these adequate?

The tests at the end of the file are all we have for now, but I had a thought today: another way to exercise this might be to artificially turn down the allowable branch range to force "real" use of islands/veneers in plausibly-large functions, e.g. maybe 1KB or so. Then we could do the usual correctness tests with our benchmarks. I'll have to think a bit about how to automate this; it ties into the idea Dan suggested before of an "evil mode", where we change some target config option(s) to make pessimistic choices everywhere (randomize block order, lower every phi as explicit moves, clobber every callee-save intentionally, etc.) and then fuzz the heck out of things. For now, perhaps I can do this manually and run the spec testsuite?
 
>     * (Islands-and-Deadlines algorithm further comment): This is clearly a complex bit of machinery.  I didn't see any single block comment explaining what the whole algorithm is.  Can you add one?  Not every detail, but at least some top level description.

Yup, added a big block comment -- thanks; more docs are always better!

>     * (revised isel driver): again, this is now more complex than it was.  Question: is lookback past block boundaries still allowed?  And how does this interact with the colouring machinery?  My impression is that lookbacks past block boundaries are still allowed, however the colours always change across block boundaries.  Hence this restricts lookbacks across boundaries to pure value trees.  So it's correct.  But if that analysis is correct, that's a non-obvious interaction that it would be good to document.

Yes, that's right; the predicate for allowing a lookback (which in practice means giving the instruction reference in the result of `get_input()`) is "same color, or producing op is pure". This is in the doc comment for `get_input()`, but perhaps I can write more high-level docs as well. I hope to write some sort of "backend writer's guide" that communicates these principles and invariants that one needs to know, eventually...

>     * (revised isel driver more): regarding lookbacks and colouring, I would like to see at least some implementation of movzx/movsx applied to loads, preferably in this patch, or at least very soon in a followup.  This is partly to improve code quality but primarily to demonstrate that the colouring infrastructure is sound (viz, to test it to some extent).

Sure; perhaps separately from this patch, as that's part of the x64 backend work (and I'm trying to touch it as little as possible in this, to avoid stepping on other ongoing work)?

>     * Is it correct to understand that the new blockorder.rs maintains the old invariant that there are no dead blocks in the output?  Or, at least, that the invariant is maintained by whatever means, that there are no dead blocks in the input to regalloc?

Yes, exactly: the output of `BlockLoweringOrder::new()` will never have dead blocks (even if the input CLIF does; the RPO naturally ensures this). Caught a few bugs with the regalloc's liveness checker so I'm somewhat confident this is the case now!

>     * I see debug! calls in potentially hottish places, eg MachBuffer::put1().  Are we sure those become zero cost in release builds?

Fixed, thanks!

**Comment by julian-seward1** [comment]
One other comment (not to block landing):

> >  (revised isel driver): again, this is now more complex than it was. Question: is lookback past block boundaries still allowed? [..]

> Yes, that's right; [..]

FTR, I am still of the opinion that lookback outside the block, even for just pure values, will turn out to be a net loss in the end, in the sense that the extra register pressure will cause much more of a loss than any minor improvements in insn selection that might result. That said, I don't have any Actual Evidence to substantiate my claim, at least currently.
